samples/go/downloader/downloader.go

```diff
@@ -0,0 +1,326 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/attic-labs/noms/go/config"
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/datas"
+	"github.com/attic-labs/noms/go/diff"
+	"github.com/attic-labs/noms/go/marshal"
+	"github.com/attic-labs/noms/go/spec"
+	"github.com/attic-labs/noms/go/types"
+	"github.com/attic-labs/noms/go/util/exit"
+	"github.com/attic-labs/noms/go/util/status"
+	"github.com/attic-labs/noms/go/util/verbose"
+	flag "github.com/juju/gnuflag"
+)
+
+var (
+	lastSourcePathFieldName = "sourcePath"
+	foundInCacheCnt         = &counter{}
+)
+
+func main() {
+	if !download() {
+		exit.Fail()
+	}
+}
+
+func usage() {
+	fmt.Println("usage: downloader --in-path <pathspec> --out-ds <dsname> --cache-ds <dsname>")
+	flag.PrintDefaults()
+}
+
+type RemoteResource struct {
+	Url string
+}
+
+type LocalResource struct {
+	Url        string
+	Downloaded bool
+	Err        string    `noms:",omitempty"`
+	BlobRef    types.Ref `noms:",omitempty"`
+}
+
+func download() (win bool) {
+	var inPathArg = flag.String("in-path", "", "root of object hierarchy to scan for RemotePhoto objects")
+	var outDsArg = flag.String("out-ds", "", "name of output dataset")
+	var cacheDsArg = flag.String("cache-ds", "", "name of photo-cache dataset")
+	var concurrencyArg = flag.Uint("concurrency", 4, "number of concurrent http calls to retrieve remote resources")
+	verbose.RegisterVerboseFlags(flag.CommandLine)
+	flag.Usage = usage
+	flag.Parse(false)
+
+	if flag.NArg() != 0 {
+		flag.Usage()
+		return
+	}
+
+	// Check that concurrency is at least 1
+	if *concurrencyArg < 1 {
+		fmt.Fprintln(os.Stderr, "error, concurrency cannot be less than 1")
+		flag.Usage()
+		return
+	}
+
+	// Check that all required args are present
+	check := []string{"in-path", *inPathArg, "out-ds", *outDsArg, "cache-ds", *cacheDsArg}
+	for i := 0; i < len(check)-1; i += 2 {
+		if check[i+1] == "" {
+			fmt.Fprintln(os.Stderr, "error: missing required argument:", check[i])
+			flag.Usage()
+			return
+		}
+	}
+
+	// Resolve the in-path arg and get the inRoot.
+	cfg := config.NewResolver()
+	db, inRoot, err := cfg.GetPath(*inPathArg)
+	if err != nil || inRoot == nil {
+		if err == nil {
+			err = errors.New("Could not find referenced value.")
+		}
+		fmt.Fprintf(os.Stderr, "Invalid input path '%s': %s\n", *inPathArg, err)
+		return
+	}
+
+	if datas.IsCommitType(inRoot.Type()) {
```

Comment:
Not having this here seems pretty bad. I'll leave it for now with the hope that this all gets standardized some day.